keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt

```diff
@@ -237,6 +195,91 @@ class ResourceActuator(
+   * we do not want to veto the artifact in the case of a bad config change or other downstream problem.
+   * Rolling back most likely will not help in these cases and it will probably cause confusion.
+   */
+  private fun handleArtifactVetoing(
+    response: VetoResponse,
+    resource: Resource<*>,
+    deliveryConfig: DeliveryConfig,
+    environment: Environment,
+    desired: Any?
+  ) {
+    if (response.vetoArtifact && resource.spec is VersionedArtifactProvider) {
+      try {
+        val versionedArtifact = desired.findArtifact()
+
+        if (versionedArtifact != null) {
+          val artifact = deliveryConfig.matchingArtifactByName(versionedArtifact.artifactName, versionedArtifact.artifactType)
+            ?: error("Artifact ${versionedArtifact.artifactType}:${versionedArtifact.artifactName} not found in delivery config ${deliveryConfig.name}")
+
+          val previousSuccess = artifactRepository.wasSuccessfullyDeployedTo(
+            deliveryConfig = deliveryConfig,
+            artifact = artifact,
+            version = versionedArtifact.artifactVersion,
+            targetEnvironment = environment.name
+          )
+
+          if (!previousSuccess) {
+            artifactRepository.markAsVetoedIn(
+              deliveryConfig = deliveryConfig,
+              veto = EnvironmentArtifactVeto(
+                reference = artifact.reference,
+                version = versionedArtifact.artifactVersion,
+                targetEnvironment = environment.name,
+                vetoedBy = "Spinnaker",
+                comment = "Automatically marked as bad because multiple deployments of this version failed and none have ever succeeded."
+              )
+            )
+            log.info(
+              "Vetoing artifact version {} of artifact {} for config {} and env {} because multiple deploys failed",
+              versionedArtifact,
+              artifact.reference + ":" + artifact.type,
+              deliveryConfig.name,
+              environment.name
+            )
+            publisher.publishEvent(ArtifactVersionVetoed(resource.application))
+          } else {
+            log.info(
+              "Not vetoing artifact version {} of artifact {} for config {} and env {} because it was previously successfully deployed",
+              versionedArtifact,
+              artifact.reference + ":" + artifact.type,
+              deliveryConfig.name,
+              environment.name
+            )
+          }
+        }
+      } catch (e: Exception) {
+        log.warn("Failed to veto presumed bad artifact version for ${resource.id}", e)
+        // TODO: emit metric
+      }
+    }
+  }
+
+  private fun Any?.findArtifact() : CompleteVersionedArtifact? =
+    when (this) {
+      is Map<*, *> -> {
+        if (this.size > 0) {
+          @Suppress("UNCHECKED_CAST")
+          (this as Map<String, VersionedArtifactProvider>).values.first()
```

Comment:
Is this to handle the case where e.g. a cluster resolves to a map of region to server group?

Reply:
This is just refactoring code that already existed, so I'm not 100% sure